README.md
# CEDViewer
[![Build Status](https://travis-ci.org/iLCSoft/CEDViewer.svg?branch=master)](https://travis-ci.org/iLCSoft/CEDViewer)

CEDViewer processor for the CED event display

CEDViewer is distributed under the [GPLv3 License](http://www.gnu.org/licenses/gpl-3.0.en.html)

[![License](https://www.gnu.org/graphics/gplv3-127x51.png)](https://www.gnu.org/licenses/gpl-3.0.en.html)

##License and Copyright
Copyright (C) 2005-2017, CED Authors

CEDViewer is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License long with this program.  If not, see <http://www.gnu.org/licenses/>.

ChangeLog
2010-11-03 18:50  engels

	* [r1805] CMakeLists.txt, ced2go/README, ced2go/ced2go,
	  ced2go/ced2go.py, src/DSTViewer.cc: merged changes from trunk
	  (r1787:r1804):
	  - renamed ced2go.py to ced2go
	  - install ced2go to bin
	  - create ced2go steering file in:
	  /tmp/ced2go_<username>_steering.xml
	  - added ced2go README

2010-10-18 14:07  engels

	* [r1785] CMakeLists.txt, ced2go/ced2go-template.xml,
	  ced2go/ced2go.py, ced2go/template.xml: merged ced2go changes from
	  trunk (1784)

2010-10-08 11:25  gaede

	* [r1767] .: ------ pre release v01-01

2010-10-08 11:24  gaede

	* [r1766] ---------------- updated version to v01-01

2010-10-06 13:20  hhoelbe

	* [r1753] add time.sleep to the mainloop of ced2go.py to avoid busy
	  waiting

2010-10-06 12:40  hhoelbe

	* [r1752] - removed the temp file ced2go/steering.xml

2010-10-06 12:39  hhoelbe

	* [r1751] - add ced2go:
	  A script which simplify the use of Marlin with CED in some cases.
	  ced2go needs the name of the LCIO file as an argument. Than it
	  look to this file find out which gearfile is needed, create an
	  steering file with this parameters. Than it starts CED and
	  Marlin.
	  
	  - release.notes updated
	  

2010-09-29 16:32  hhoelbe

	* [r1741] removed debug statement

2010-09-29 14:21  hhoelbe

	* [r1738] CEDViewer: Helix part of tracks was not pickable. Fixed

2010-09-29 13:59  hhoelbe

	* [r1737] CEDViewer now displays also Reconstructed Particles. If
	  there are no hits, than he draw the helix.

2010-09-22 16:00  gaede

	* [r1728] added more recent example (ILD_00)

2010-09-22 15:59  gaede

	* [r1727] changed default layer to more reasonable values; fixed
	  printout for default layers

2010-09-22 14:53  gaede

	* [r1725] reversed order of marker and size in steering to
	  original: colName, marker, size

2010-09-22 14:37  gaede

	* [r1724] fixed drawing of tracks w/ reference points (need to set
	  compile time flag, as old rtacks have worng reference point
	  (PCA))

2010-09-15 15:41  hhoelbe

	* [r1714] the viewers does not reset the layer descriptions anymore
	  at startup

2010-09-09 13:04  hhoelbe

	* [r1704] - now used MarlinCED::add_layer_description to add
	  multiples descriptions
	  for one layer

2010-08-25 15:26  hhoelbe

	* [r1685] Cones have got a central line for picking

2010-08-20 14:34  gaede

	* [r1682] changed layer description to collection name

2010-08-18 16:45  hhoelbe

	* [r1678] DSTViewer and CEDViewer add their layer descriptions to
	  CED menu and help

2010-08-11 16:19  hhoelbe

	* [r1673] rollback to last revision

2010-08-11 15:44  hhoelbe

	* [r1672] cedviewer now add his layer description to ced (for menu
	  & help frame)

2010-07-02 11:28  /O=GermanGrid/OU=DESY/CN=Jan Engels

	* [r1641] ---------- v01-00 ----------

2010-06-29 08:45  gaede

	* [r1621] turned -32 compatibility mode OFF

2010-06-29 08:43  gaede

	* [r1620] turned -32 bit mode OFF

2010-06-23 17:03  hhoelbe

	* [r1611] fixed id

2010-06-23 15:21  /O=GermanGrid/OU=DESY/CN=Jan Engels

	* [r1602] added exclude pattern for .svn dirs

2010-06-23 14:42  hhoelbe

	* [r1601] make genericviewer ready for picking

2010-06-23 09:55  engels

	* [r1594] changed to version v01-00-00

2010-06-18 15:38  hhoelbe

	* [r1591] Enabled picking (hauke)

2010-06-17 14:43  aplin

	* [r1586] added ctest

2010-06-17 07:36  /O=GermanGrid/OU=DESY/CN=Jan Engels

	* [r1581] removed compiler warnings
	  bug fix: assignment instead of comparison inside if construct

2010-06-11 10:18  gaede

	* [r1578] changed to version v01-00

2010-06-11 10:09  gaede

	* [r1577] added Picking description (H.Hoelbe)

2010-06-11 10:04  gaede

	* [r1576] some cleanup of code and fix for drawing MCParticles

2010-06-10 14:24  gaede

	* [r1575] typo ...

2010-06-10 14:23  gaede

	* [r1574] type in printout

2010-06-10 14:05  gaede

	* [r1573] add some printout if collection not found and layer
	  number used for drawing

2010-06-10 14:04  gaede

	* [r1572] add streamlog for APPLE

2010-05-27 13:55  engels

	* [r1563] merged branch: hauke_dev (TODO: change this log msg)

2010-02-09 14:42  engels

	* [r1466] ----------- v00-07-02 ----------

2010-02-07 11:17  gaede

	* [r1458] Added new parameter 'DrawInLayer' to allow the user to
	  specify the CED layer number

2009-11-25 16:31  /C=DE/O=GermanGrid/OU=DESY/CN=Steven Aplin

	* [r1450] ---------------- release v00-07-01 ----------------

2009-09-08 13:04  stewartm

	* [r1365] SM-H: Removed pi = acos(-1)

2008-10-23 15:59  engels

	* [r1258] --------- v00-07 ---------

2008-10-15 08:08  engels

	* [r1248] added -Wno-long-long definition

2008-10-14 11:02  engels

	* [r1243] added 32 bit compatibility option

2008-10-14 10:58  engels

	* [r1242] removed some compiler warnings

2008-09-15 10:13  darasz

	* [r1234] modified drawing of clusters to ellipsoids and tidyed up
	  color map code

2008-09-11 11:42  engels

	* [r1232] small fix for compiling with gcc 3.2

2008-09-09 14:42  darasz

	* [r1231] updated doc

2008-09-09 14:38  darasz

	* [r1230] implemented a DSTViewer: draws ReconstructedParticles as
	  helices and straight lines; Clusters as Cylinders (Ellipsoids)
	  and different jet hypotheses as cones

2008-07-23 14:57  gaede

	* [r1198] ------------- version v00-06 --------------------

2008-07-23 14:56  gaede

	* [r1197] made dependant on CED directly

2008-07-03 08:55  rasp

	* [r1122] New version of the VertexViewer processor.

2008-07-03 08:55  rasp

	* [r1121] New version of VertexViewer processor. Bug related to the
	  calculation of the global rotation angle Ph0 for the ladders is
	  fixed

2008-05-28 14:47  engels

	* [r1035] patched to make gcc 4.3 compliant

2008-04-16 15:28  gaede

	* [r953] -------------- version v00-05 ---------------

2008-04-16 15:11  gaede

	* [r952] initial version of VertexViewer

2008-04-16 15:08  gaede

	* [r951] replaced std::cout with streamlog_out( DEBUG )

2008-01-22 08:52  engels

	* [r862] set both variables INCLUDE_DIR and INCLUDE_DIRS for
	  compatibility


ced2go/ced2go.py
#!/usr/bin/python
import sys
import re
import subprocess
import os
import signal
import time
import getopt


#Help output
def help():
    print "Usage: " + str(sys.argv[0]) + " [OPTION]  <LCIO Input File>"
    print "Options: "
    print "     -v            Viewer1"
    print "     -v            Viewer2"
    print "     -d            Gearfile"
    print "     -h            Help"
    print "All options are optional"
    print "Viewers are e.g. CEDViewer, DSTViewer, GenericViewer"

#import shlex
#from Tkinter import *
#import threading

#The global variables, paths and commands
viewer   = "CEDViewer" #default viewer

#templateXMLFile = "/afs/desy.de/user/h/hhoelbe/workspace/ilcsoft/v01-07/Wrapper/template.xml"
templateXMLFile = os.path.join( os.path.abspath(sys.path[0]), "ced2go-template.xml" )


if not os.path.isfile(templateXMLFile):
    print "Error: Template File \"" + templateXMLFile + "\" not found!"
    sys.exit()


#extractDetector="/afs/desy.de/user/h/hhoelbe/workspace/ilcsoft/v01-07/Wrapper/build/extractdetector"
extractDetector = os.path.join( os.path.abspath(sys.path[0]), "extractdetector" )

if not os.path.isfile(extractDetector):
    print "Error: Helper Tool \"" + extractDetector + "\" not found!"
    sys.exit()

steeringFile="/tmp/ced2go_steering.xml"
CED= "glced"
Marlin="Marlin"

try:
    standartconfig=os.environ["STANDARDCONFIG"]
except: 
    print "Error: Please source the init ilcsoft script first"
    sys.exit(1)

gearfiles={'ILD_00':standartconfig + '/mc2008/gear_ILD_00.xml',
        'LDC01_06Sc':standartconfig+ '/mc2008/gear_LDC01_06Sc.xml',
        'LDCPrime_02Sc':standartconfig+'/mc2008/gear_LDCPrime_02Sc.xml',
        'LDC_GLD_01Sc':standartconfig+ '/mc2008/gear_LDC_GLD_01Sc.xml'
}

keys = {'$LCIOInputFiles$':'', 
        '$GearXMLFile$':'',
        '$Viewer$':''
       }

#Parse the command line arguments
try: 
    opts, args = getopt.getopt(sys.argv[1:], "hv:d:", "help") 
    #print 'Opts:',opts
    #print 'Extra parameters:',args
except getopt.GetoptError, err:
    print str(err)
    help()
    sys.exit(2)

detector = ""
for key in opts:
    #print key[0] + str(key[1])
    if key[0] == "-d":
        detector=key[1]
        if not os.path.isfile(detector):
            print "Error: Gearfile \"" + detector + "\" not found!"
            sys.exit()
    if key[0]  == "-v":
          keys['$Viewer$'] += "\n<processor name=\"My" + str(key[1]) + "\"/>"
    if key[0] in ("-h", "--help"):
        help()
        sys.exit(0)

if keys['$Viewer$'] == '': #default
    keys['$Viewer$'] = "<processor name=\"My" + viewer + "\"/>" 


#class Thread1(threading.Thread):
#    def run(self):
#        print "thread 1 startet and block"
#        while True:
#            pass

#Thread1().start()
#print "main sleep 5sec"
#time.sleep(5)
#sys.exit()

#check if lcio file is given
if not len(args) == 1:
    help()
    sys.exit()


#check if lcio file exists
if not os.path.isfile(args[0]):
    print "Error: LCIO File \"" + sys.argv[1] + "\" not found!"
    help()
    sys.exit()
    
keys['$LCIOInputFiles$']=args[0]



if detector != "": #user defined gearfile
     keys['$GearXMLFile$'] = detector
else:  #find out the detector name used in the lcio file
    detector= str(subprocess.Popen([extractDetector,keys['$LCIOInputFiles$']], stdout=subprocess.PIPE).communicate()[0])
    detector=detector.rstrip() #remove newline

    #print '"' + detector + '"'

    tmp = detector.split("\n")
    if len(tmp) != 1:
        print "Warning: LCIO File contains more than one detector model, try the first one"
        detector=tmp[0]
    try: 
        keys['$GearXMLFile$'] = gearfiles[detector] 
    except(KeyError):
        print "Error: Unknown detector in LCIO file"
        sys.exit(1)

#root = Tk()

#w=Label(root, text="hello world")
#w.pack()

#root.mainloop()


#replace the words from dict with the keywords in the template file
templateFH = open(templateXMLFile,'r')
xmlFH = open(steeringFile,'w')

for line in templateFH:
    for key in keys: 
        line= re.sub(re.escape(key),keys[key],line);
    xmlFH.write(line) 
xmlFH.close()
templateFH.close()

#start ced and marlin
CEDobj = subprocess.Popen([CED,'-geometry', '600x400'])
Marlinobj = subprocess.Popen([Marlin, steeringFile])

#terminate the both processes when one is closed
while True: 
    if Marlinobj.poll() != None:
        os.kill(CEDobj.pid,15)
        break

    if CEDobj.poll() != None: 
        os.kill(Marlinobj.pid,15)
        break

    time.sleep(0.5) #sleep for 0.5 seconds

print "Done"
